### PR TITLE
Bug 2254547: object: improve the error handling for multisite objs

### PR DESCRIPTION
**Pull request by parth-gr**
here is the https://go.dev/play/p/SS9Q-dAiIx3 example which says the error handling was wrongly implemented

Signed-off-by: parth-gr <paarora@redhat.com>
(cherry picked from commit 46c241433d8722c911bff953dbe8c6c302240a95)

**Description of your changes:**

**Which issue is resolved by this Pull Request:**
Resolves #

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure)).
- [ ] **Skip Tests for Docs**: If this is only a documentation change, add the label `skip-ci` on the PR.
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

**Comment by travisn** [comment]
This test is failing in the object suite: 
```
--- FAIL: TestCephObjectSuite/TestWithTLS/run_a_second_object_store/create_CephObjectStore (400.56s)
```
I seem to recall we didn't backport this change to 1.12 upstream because of this error. Any insight to the failure?

**Comment by parth-gr** [comment]
```
runner@fv-az1204-469:~/work/rook/rook$ kubectl describe cephobjectstore -nobject-ns other-tls-test-store 
Name:         other-tls-test-store
Namespace:    object-ns
Labels:       <none>
Annotations:  <none>
API Version:  ceph.rook.io/v1
Kind:         CephObjectStore
Metadata:
  Creation Timestamp:  2024-01-30T15:37:20Z
  Finalizers:
    cephobjectstore.ceph.rook.io
  Generation:        2
  Resource Version:  6306
  UID:               1f190946-165b-41db-9b51-bf13c0a8e99e
Spec:
  Data Pool:
    Erasure Coded:
      Coding Chunks:  0
      Data Chunks:    0
    Mirroring:
    Quotas:
    Replicated:
      Size:  1
    Status Check:
      Mirror:
  Gateway:
    Instances:  1
    Placement:
    Resources:
    Secure Port:          80
    Ssl Certificate Ref:  other-tls-test-store
  Health Check:
  Metadata Pool:
    Compression Mode:  passive
    Erasure Coded:
      Coding Chunks:  0
      Data Chunks:    0
    Mirroring:
    Quotas:
    Replicated:
      Size:  1
    Status Check:
      Mirror:
  Zone:
    Name:  
Status:
  Endpoints:
    Insecure:
    Secure:
      https://rook-ceph-rgw-other-tls-test-store.object-ns.svc:80
  Phase:  Failure
Events:
  Type     Reason           Age                 From                         Message
  ----     ------           ----                ----                         -------
  Warning  ReconcileFailed  16m (x19 over 41m)  rook-ceph-object-controller  failed to reconcile CephObjectStore "object-ns/other-tls-test-store". failed to create object store deployments: failed to configure multisite for object store: failed create ceph multisite for object-store ["other-tls-test-store"]: failed to create ceph ["zone" "--rgw-zone=other-tls-test-store" "unable to initialize zonegroup other-tls-test-store: (2) No such file or directory. "] %!q(MISSING), for reason %!q(MISSING): exit status 2
runner@fv-az1204-469:~/work/rook/rook$ 
```


I see this, which the original issue why the BZ was reported

**Comment by parth-gr** [comment]
It is because the zonegroup is not created

```
sh-4.4$ radosgw-admin zonegroup get --rgw-zonegroup=other-tls-test-store
failed to load zonegroup: (2) No such file or directory
```

**Comment by travisn** [comment]
As discussed offline, the realm arg isn't being passed for the `zonegroup get`.

File: pkg/operator/ceph/object/objectstore.go [hunk 1]
```diff
@@ -392,55 +392,61 @@ func getZoneEndpoints(objContext *Context, serviceEndpoint string) ([]string, bo
...
-			logger.Debugf("created realm %q", objContext.Realm)
-		} else {
-			return errorOrIsNotFound(err, "'radosgw-admin realm get' failed with code %d, for reason %q. %v", strconv.Itoa(code), output, string(kerrors.ReasonForError(err)))
-		}
+		return err
 	}
 
-	// create the zonegroup if it doesn't exist yet
-	output, err = RunAdminCommandNoMultisite(objContext, true, "zonegroup", "get", realmArg, zoneGroupArg)
+	err = createMultisiteConfigurations(objContext, "zonegroup", realmArg+zoneGroupArg, "create", "--master", realmArg, endpointArg)
```
Comment (parth-gr) [inline pkg/operator/ceph/object/objectstore.go:442]: updated

File: pkg/operator/ceph/object/objectstore.go [hunk 2]
```diff
@@ -392,55 +392,66 @@ func getZoneEndpoints(objContext *Context, serviceEndpoint string) ([]string, bo
 	return zoneEndpointsList, isEndpointAlreadyExists, nil
 }
 
+func createMultisiteConfigurations(objContext *Context, configType, configTypeArg string, args ...string) error {
+	args = append([]string{configType}, args...)
+	args = append(args, configTypeArg)
+	// get the multisite config before creating
+	configTypeArgs := []string{configType, "get", configTypeArg}
+	if configType == "zonegroup" {
```
Comment (parth-gr) [inline pkg/operator/ceph/object/objectstore.go:400]: Added a check

**Comment by travisn** [comment]
It seems the zonegroup is still failing. From the operator log:
```
2024-01-30 20:21:53.033880 D | ceph-object-controller: creating realm, zone group, zone for object-store other-tls-test-store
2024-01-30 20:21:53.033930 D | exec: Running command: radosgw-admin realm get --rgw-realm=other-tls-test-store --cluster=object-ns --conf=/var/lib/rook/object-ns/object-ns.config --name=client.admin --keyring=/var/lib/rook/object-ns/client.admin.keyring
2024-01-30 20:21:53.148372 D | exec: Running command: radosgw-admin realm create --rgw-realm=other-tls-test-store --cluster=object-ns --conf=/var/lib/rook/object-ns/object-ns.config --name=client.admin --keyring=/var/lib/rook/object-ns/client.admin.keyring
2024-01-30 20:21:53.251465 D | ceph-object-controller: created "realm" "--rgw-realm=other-tls-test-store"
2024-01-30 20:21:53.251507 D | exec: Running command: radosgw-admin zonegroup get --master --rgw-zonegroup=other-tls-test-store --cluster=object-ns --conf=/var/lib/rook/object-ns/object-ns.config --name=client.admin --keyring=/var/lib/rook/object-ns/client.admin.keyring
2024-01-30 20:21:53.351694 D | op-osd: checking osd processes status.
2024-01-30 20:21:53.351732 D | exec: Running command: ceph osd dump --connect-timeout=15 --cluster=object-ns --conf=/var/lib/rook/object-ns/object-ns.config --name=client.admin --keyring=/var/lib/rook/object-ns/client.admin.keyring --format json
2024-01-30 20:21:53.552867 D | exec: Running command: radosgw-admin zone get --rgw-zone=other-tls-test-store --cluster=object-ns --conf=/var/lib/rook/object-ns/object-ns.config --name=client.admin --keyring=/var/lib/rook/object-ns/client.admin.keyring
2024-01-30 20:21:53.658482 D | exec: Running command: radosgw-admin zone create --master --endpoints=https://rook-ceph-rgw-other-tls-test-store.object-ns.svc:80 --rgw-realm=other-tls-test-store --rgw-zonegroup=other-tls-test-store --rgw-zone=other-tls-test-store --cluster=object-ns --conf=/var/lib/rook/object-ns/object-ns.config --name=client.admin --keyring=/var/lib/rook/object-ns/client.admin.keyring
2024-01-30 20:21:53.800493 D | ceph-object-controller: object store "object-ns/other-tls-test-store" status updated to "Failure"
2024-01-30 20:21:53.800637 D | clusterdisruption-controller: reconciling "object-ns/"
2024-01-30 20:21:53.800768 D | ceph-spec: update event from a CR: "other-tls-test-store"
2024-01-30 20:21:53.800808 D | ceph-spec: update event on CephObjectStore CR
2024-01-30 20:21:53.800906 D | ceph-spec: update event from a CR: "other-tls-test-store"
2024-01-30 20:21:53.800921 D | ceph-spec: update event on CephObjectStore CR
2024-01-30 20:21:53.801095 D | clusterdisruption-controller: could not match failure domain. defaulting to "host"
2024-01-30 20:21:53.808505 D | op-k8sutil: kubernetes version fetched 1.27.2
2024-01-30 20:21:53.810586 D | op-k8sutil: kubernetes version fetched 1.27.2
2024-01-30 20:21:53.810716 D | exec: Running command: ceph status --connect-timeout=15 --cluster=object-ns --conf=/var/lib/rook/object-ns/object-ns.config --name=client.admin --keyring=/var/lib/rook/object-ns/client.admin.keyring --format json
2024-01-30 20:21:53.822881 D | ceph-object-controller: object store "object-ns/other-tls-test-store" status updated to "Failure"
2024-01-30 20:21:53.822971 E | ceph-object-controller: failed to reconcile CephObjectStore "object-ns/other-tls-test-store". failed to create object store deployments: failed to configure multisite for object store: failed create ceph multisite for object-store ["other-tls-test-store"]: failed to create ceph ["zone" "--rgw-zone=other-tls-test-store" "unable to initialize zonegroup other-tls-test-store: (2) No such file or directory. "] %!!(MISSING)q(MISSING), for reason %!!(MISSING)q(MISSING): exit status 2
```

**Comment by parth-gr** [comment]
Got fixed

**Review by travisn — APPROVED** [review]
/approve
/lgtm

**Pull request merged**